javalib/src/main/scala/java/util/AbstractList.scala

```diff
@@ -1,3 +1,293 @@
 package java.util
 
-abstract class AbstractList[E] extends AbstractCollection[E] with List[E]
+import scala.collection.JavaConversions._
+
+import scala.annotation.tailrec
+
+abstract class AbstractList[E] protected () extends AbstractCollection[E] with List[E] {
+  self =>
+
+  override def add(element: E): Boolean = {
+    add(size, element)
+    true
+  }
+
+  def add(index: Int, element: E): Unit =
+    throw new UnsupportedOperationException
+
+  def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
+    checkIndexOnBounds(index)
+    for ((elem, i) <- c.iterator().zipWithIndex)
+      add(index + i, elem)
+    c.nonEmpty
+  }
+
+  override def clear(): Unit =
+    removeRange(0, size)
+
+  override def equals(o: Any): Boolean = {
+    if (o.asInstanceOf[AnyRef] eq this)
+      return true
+
+    o match {
+      case other: List[_] =>
+        val (e1, e2) = (this.listIterator, other.listIterator)
+        while (e1.hasNext && e2.hasNext) {
+          val (o1, o2 )= (e1.next(), e2.next())
+          if (!(o1 == null && o2 == null) && o1 != o2)
+            return false
+        }
+        !e1.hasNext && !e2.hasNext
```

Comment:
that was also my first approach, but then I realised that the lists might have to compute their size and hence this would imply two traversals of the list in some cases.